main.go

```diff
@@ -0,0 +1,313 @@
+			candidates = append(candidates, appIconCandidate{
+				FileName: fileInfo.Name(),
+				FileSize: fileInfo.Size(),
+			})
+		}
+	}
+	candidateURLs, err := getUploadURL(iconCandidateURL, buildTriggerToken, candidates)
+	if err != nil {
+		return fmt.Errorf("failed to get candidate target URLs, error: %s", err)
+	}
+
+	for _, candidateURL := range candidateURLs {
+		err := uploadIcon(iconsDir, candidateURL)
+		if err != nil {
+			return fmt.Errorf("failed to upload icon, error: %s", err)
+		}
+	}
+
+	log.Donef("submitted")
+	return nil
+}
+
+func getUploadURL(url string, buildTriggerToken string, appIcons []appIconCandidate) ([]appIconCandidateURL, error) {
+	var uploadURLs []appIconCandidateURL
+	if err := retry.Times(3).Wait(5 * time.Second).Try(func(attempt uint) error {
+		if attempt > 0 {
+			log.Warnf("%d query attempt failed", attempt)
+		}
+
+		data, err := json.Marshal(appIcons)
+		if err != nil {
+			return fmt.Errorf("failed to marshal json")
+		}
+
+		request, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(data))
+		if err != nil {
+			return fmt.Errorf("failed to create request")
+		}
+		request.Header.Set("Authorization", fmt.Sprintf("token %s", buildTriggerToken))
+		request.Header.Set("Content-Type", "application/json")
+
+		resp, err := http.DefaultClient.Do(request)
+		if err != nil {
+			return fmt.Errorf("failed to submit, error: %s", err)
+		}
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.Errorf("Failed to close response body, error: %s", err)
+			}
+		}()
+
+		if err != nil {
+			return fmt.Errorf("failed to submit, err: %s", err)
+		}
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read respnse body, error: %s", err)
+		}
+
+		if resp.StatusCode != http.StatusCreated {
+			return fmt.Errorf("invalid status code: %d, headers: %s, body: %s", resp.StatusCode, resp.Header, body)
+		}
+
+		decoded := map[string][]appIconCandidateURL{
+			"data": uploadURLs,
+		}
+
+		err = json.Unmarshal(body, &decoded)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal resoponse bodys")
+		}
+		uploadURLs = decoded["data"]
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to upload, error: %s", err)
+	}
+	return uploadURLs, nil
+}
+
+func uploadIcon(basePath string, iconCandidate appIconCandidateURL) error {
+	if err := retry.Times(3).Wait(5 * time.Second).Try(func(attemp uint) error {
+		if attemp != 0 {
+			log.Warnf("%d query attemp failed", attemp)
+		}
+
+		file, err := os.Open(path.Join(basePath, iconCandidate.FileName))
+		if err != nil {
+			return fmt.Errorf("failed to open file")
+		}
+		defer func() {
+			if err := file.Close(); err != nil {
+				log.Warnf("failed to close file")
+			}
+		}()
+
+		data, err := ioutil.ReadAll(file)
+		if err != nil {
+			return fmt.Errorf("can not read file")
+		}
+		if int64(len(data)) != iconCandidate.FileSize {
+			return fmt.Errorf("content-lenght has to match signed URL")
```

Comment:
this error message is strange a bit, other than that is it possible that file.Size does not match to the fileinfo.Size?

Reply:
It should match (if the files on disk do not change in that small time, other than that is serves as a comment that it is important to match the Content-lenght of the signed s3 URL. If reading from a stream rather than byte array, it is not set by the golang http.NewRequest.